src/encoder/encoderopus.cpp

```diff
@@ -0,0 +1,439 @@
+    (void)errorMessage;
+    int result;
+
+    if (samplerate != 48000) {
+        kLogger.warning() << "initEncoder failed: samplerate not supported by Opus";
+
+        QString invalidSamplerate = QObject::tr(
+                "Using Opus at samplerates other than 48 kHz "
+                "is not supported by the Opus encoder. Please use "
+                "48000 Hz in \"Sound Hardware\" preferences "
+                "or switch to a different encoding.");
+
+        ErrorDialogProperties* props = ErrorDialogHandler::instance()->newDialogProperties();
+        props->setType(DLG_WARNING);
+        props->setTitle(QObject::tr("Encoder"));
+        props->setText(invalidSamplerate);
+        props->setKey(invalidSamplerate);
+        ErrorDialogHandler::instance()->requestErrorDialog(props);
+        return -1;
+    }
+
+    m_samplerate = samplerate;
+    m_pOpus = opus_encoder_create(m_samplerate, m_channels, OPUS_APPLICATION_AUDIO, &result);
+
+    if (result != OPUS_OK) {
+        kLogger.warning() << "opus_encoder_create failed:" << opusErrorString(result);
+        return -1;
+    }
+
+    // Optimize encoding for high-quality music
+    opus_encoder_ctl(m_pOpus, OPUS_SET_COMPLEXITY(10)); // Highest setting
+    opus_encoder_ctl(m_pOpus, OPUS_SET_SIGNAL(OPUS_SIGNAL_MUSIC));
+
+    if(m_bitrateMode == 0) {
+        // == Constrained VBR ==
+        // Default mode, gives the best quality/bitrate compromise
+        opus_encoder_ctl(m_pOpus, OPUS_SET_BITRATE(m_bitrate * 1000)); // convert to bits/second
+        opus_encoder_ctl(m_pOpus, OPUS_SET_VBR(1)); // default value in libopus
+        opus_encoder_ctl(m_pOpus, OPUS_SET_VBR_CONSTRAINT(1)); // Constrained VBR
+    } else if(m_bitrateMode == 1) {
+        // == CBR (quality loss at low bitrates) ==
+        opus_encoder_ctl(m_pOpus, OPUS_SET_BITRATE(m_bitrate * 1000)); // convert to bits/second
+        opus_encoder_ctl(m_pOpus, OPUS_SET_VBR(0));
+    } else if(m_bitrateMode == 2) {
+        // == Full VBR ==
+        opus_encoder_ctl(m_pOpus, OPUS_SET_VBR(1));
+        opus_encoder_ctl(m_pOpus, OPUS_SET_VBR_CONSTRAINT(0)); // Unconstrained VBR
+    }
+
+    // Size the input FIFO buffer with twice the maximum possible sample count that can be
+    // processed at once, to avoid skipping frames or waiting for the required sample count
+    // and encode at a regular pace.
+    // This is set to the buffer size of the sidechain engine because
+    // Recording (which uses this engine) sends more samples at once to the encoder than
+    // the Live Broadcasting implementation
+    m_pFifoBuffer = new FIFO<CSAMPLE>(EngineSideChain::SIDECHAIN_BUFFER_SIZE * 2);
+
+    double samplingPeriodMs = (1.0/((float)m_samplerate)) * 1000.0;
+    double samplesPerChannel = kOpusFrameMs / samplingPeriodMs;
+
+    m_readRequired = samplesPerChannel * m_channels;
+    m_pFifoChunkBuffer = new CSAMPLE[m_readRequired * sizeof(CSAMPLE)]();
```

Comment:
Please use SampleBuffer instead of doing your own memory management. This would have prevented the error to over-allocate memory by multiplying with sizeof(CSAMPLE)